polyfill/lib/calendar.mjs

```diff
@@ -333,53 +332,1354 @@ function resolveNonLunisolarMonth(calendarDate) {
+const makeHelperGregorian = (id, originalEras) => {
+  const { eras, anchorEra } = adjustEras(originalEras);
+  return ObjectAssign({}, nonIsoHelperBase, {
+    id,
+    eras,
+    anchorEra,
+    calendarType: 'solar',
+    inLeapYear(calendarDate /*, cache */) {
+      const { year } = this.estimateIsoDate(calendarDate);
+      return isGregorianLeapYear(year);
+    },
+    monthsInYear(/* calendarDate */) {
+      return 12;
+    },
+    minimumMonthLength(calendarDate) {
+      const { month } = calendarDate;
+      if (month === 2) return this.inLeapYear(calendarDate) ? 29 : 28;
+      return [4, 6, 9, 11].indexOf(month) >= 0 ? 30 : 31;
+    },
+    maximumMonthLength(calendarDate) {
+      return this.minimumMonthLength(calendarDate);
+    },
+    /** Fill in missing parts of the (year, era, eraYear) tuple */
+    completeEraYear(calendarDate) {
+      const checkField = (name, value) => {
+        const currentValue = calendarDate[name];
+        if (currentValue != null && currentValue != value) {
+          throw new RangeError(`Input ${name} ${currentValue} doesn't match calculated value ${value}`);
+        }
+      };
+      const eraFromYear = (year) => {
+        let eraYear;
+        const adjustedCalendarDate = { ...calendarDate, year };
+        const matchingEra = this.eras.find((e, i) => {
+          if (i === this.eras.length - 1) {
+            if (e.reverseOf) {
+              // This is a reverse-sign era (like BCE) which must be the oldest
+              // era. Count years backwards.
+              if (year > 0) throw new RangeError(`Signed year ${year} is invalid for era ${e.name}`);
+              eraYear = e.anchorEpoch.year - year;
+              return true;
+            }
+            // last era always gets all "leftover" (older than epoch) years,
+            // so no need for a comparison like below.
+            eraYear = year - e.anchorEpoch.year + (e.hasYearZero ? 0 : 1);
+            return true;
+          }
+          const comparison = nonIsoHelperBase.compareCalendarDates(adjustedCalendarDate, e.anchorEpoch);
+          if (comparison >= 0) {
+            eraYear = year - e.anchorEpoch.year + (e.hasYearZero ? 0 : 1);
+            return true;
+          }
+          return false;
+        });
+        if (!matchingEra) throw new RangeError(`Year ${year} was not matched by any era`);
+        return { eraYear, era: matchingEra.name };
+      };
+
+      let { year, eraYear, era } = calendarDate;
+      if (year != null) {
+        ({ eraYear, era } = eraFromYear(year));
+        checkField('era', era);
+        checkField('eraYear', eraYear);
+      } else if (eraYear != null) {
+        const matchingEra =
+          era === undefined ? undefined : this.eras.find((e) => e.name === era || e.genericName === era);
+        if (!matchingEra) throw new RangeError(`Era ${era} (ISO year ${eraYear}) was not matched by any era`);
+        if (eraYear < 1 && matchingEra.reverseOf) {
```

Comment:
The convention I used (at least I think I did!) was the same as what legacy Date does with era-less input: out of range values are allowed and evaluated relative to the epoch of the era. So `era: 'ce', eraYear: 0` is treated as ISO year 0 and will be stored as 1 BCE. An exception is reverse eras, which I require to be positive because -1 BCE is clearly a bug. 

We could also choose to reject 0 CE inputs. I don't feel strongly either way.

Reply:
Hm, I think if we're in a 'constrain' context that makes sense, but a calendar should reject an eraYear that's not valid for the context of a given era in a 'restrict' context. It seems inconsistent to coerce eraYear 0 to CE 1 otherwise.

Reply:
Sorry, I wasn't clear. The current behavior for all eras is *not* to constrain `eraYear`, but rather to evaluate `eraYear` as a signed value in context of the era's epoch. So 0 CE => 1 BCE, and Meiji 100 => Showa 42.

That's why I'm a little skeptical about constrain behavior because it's a little ambiguous about what it should do: constrain to first/last day or era?  Or current behavior?

Reply:
BTW, one reason why the current behavior is as-is is to make it cleaner to deal with dates in years where eras end mid-year. For example, Jan 1 Reiwa 1 is technically in the previous era because Reiwa started May 1. By always treating eraYear as a signed value without constraints, it includes handling the Jan 1 Reiwa 1 case too.  I think there was discussion of this in another issue but forget which.

Reply:
Hm, I think this is tricky. There's no clear *temporal* spec for this, but it does seem in the spirit of constrain vs reject that Meiji 100 would be a RangeError when treated strictly, and Showa 42 when treated loosely. Same for 0 CE vs 0 BCE, for example. My preference would be not to evaluate `eraYear` as a signed value in the context of an epoch; that's what `year` is for. This could lead to developer mistakes, too, though it's admittedly a little obscure. Let me see if I can find that other discussion after reviewing the tests, before I suggest any changes.

Reply:
> I'm so sorry to ask a little more of you here, but... could you add just a few test cases for `Temporal.PlainMonthDay.from` that cover #1239 / #1316 ? I think that'd be enough tests for now. Thank you so much again for doing all of this work, sorry to pick over it with such a fine-tooth comb.

No worries, this is a great review. I'll add these tests tomorrow.

Reply:
If OK with you, I'd prefer to handle this question as a follow-up PR, because this case seems ambiguous enough that we may want to discuss it with a wider group before committing to a course of action. To summarize, there are a few questions to answer: 

1. How to handle out-of-bounds eraYear with `reject`?
2. How to handle out-of-bounds eraYear with `constrain`?
3. How to handle out-of-bounds months/days (but in-bounds eraYear) with `reject`?
4. How to handle out-of-bounds months/days (but in-bounds eraYear) with `constrain`?

AFAIK, we already decided (3) and (4) in #1300, which was to accept those values without exceptions. So unless there's new information I'd prefer to leave those closed decisions closed.